File: crates/merchant-core/src/economics.rs
use rust_decimal::{Decimal, RoundingStrategy};
use serde::{Deserialize, Serialize};

use crate::{CostAssumptions, DecimalString, DomainError};

#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub enum ScenarioName { Low, Base, High }

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct EconomicsScenario {
    pub scenario: ScenarioName,
    pub selling_price: DecimalString,
    pub acquisition_cost: DecimalString,
    pub shipping_cost: DecimalString,
    pub marketplace_fee_rate: DecimalString,
    pub marketplace_fee: DecimalString,
    pub payment_fee_rate: DecimalString,
    pub payment_fee: DecimalString,
    pub other_costs: DecimalString,
    pub total_cost: DecimalString,
    pub gross_profit: DecimalString,
    pub gross_margin_percent: DecimalString,
}

pub fn calculate_scenarios(assumptions: &CostAssumptions) -> Result<Vec<EconomicsScenario>, DomainError> {
    [ (ScenarioName::Low, assumptions.scenario_prices.low), (ScenarioName::Base, assumptions.scenario_prices.base), (ScenarioName::High, assumptions.scenario_prices.high) ]
        .into_iter()
        .map(|(scenario, price)| calculate_scenario(scenario, price.ok_or(DomainError::MissingScenarioPrice)?, assumptions))
        .collect()
}

fn calculate_scenario(scenario: ScenarioName, selling_price: DecimalString, assumptions: &CostAssumptions) -> Result<EconomicsScenario, DomainError> {
    let selling = selling_price.decimal();
    if selling <= Decimal::ZERO { return Err(DomainError::SellingPriceMustBePositive); }
    let marketplace_fee = selling * assumptions.marketplace_fee_rate.decimal() / Decimal::from(100);
    let payment_fee = selling * assumptions.payment_fee_rate.decimal() / Decimal::from(100);
    let total_cost = assumptions.acquisition_cost.decimal() + assumptions.shipping_cost.decimal() + marketplace_fee + payment_fee + assumptions.other_costs.decimal();
    let gross_profit = selling - total_cost;
    let gross_margin_percent = gross_profit / selling * Decimal::from(100);
    Ok(EconomicsScenario { scenario, selling_price, acquisition_cost: assumptions.acquisition_cost, shipping_cost: assumptions.shipping_cost, marketplace_fee_rate: assumptions.marketplace_fee_rate, marketplace_fee: rounded(marketplace_fee), payment_fee_rate: assumptions.payment_fee_rate, payment_fee: rounded(payment_fee), other_costs: assumptions.other_costs, total_cost: rounded(total_cost), gross_profit: rounded(gross_profit), gross_margin_percent: rounded(gross_margin_percent) })
}

fn rounded(value: Decimal) -> DecimalString { DecimalString::from_decimal(value.round_dp_with_strategy(2, RoundingStrategy::MidpointAwayFromZero)) }

File: packages/core/src/economics.ts
import type {
  CostAssumptions,
  EconomicsScenario,
  ScenarioName,
} from "@open-merchant/shared";

import { MoneyError, formatAmount, parseAmount, roundAmount } from "./money";

/**
 * Deterministic unit-economics engine. Ported from the legacy Rust
 * `merchant-core::economics` with byte-identical outputs (see golden
 * fixtures): fees enter totals UNROUNDED and every emitted field is rounded
 * independently, half-away-from-zero, to two decimal places.
 */

export class DomainError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "DomainError";
  }
}

export const MISSING_SCENARIO_PRICE = "Every selling-price scenario must be provided";
export const SELLING_PRICE_POSITIVE = "Selling prices must be positive";

const SCENARIO_ORDER: readonly (readonly [ScenarioName, keyof CostAssumptions["scenarioPrices"]])[] = [
  ["low", "low"],
  ["base", "base"],
  ["high", "high"],
];

export function calculateScenario(
  scenarioName: ScenarioName,
  sellingPriceRaw: string | null,
  assumptions: CostAssumptions,
): EconomicsScenario {
  if (sellingPriceRaw === null || sellingPriceRaw === undefined) {
    throw new DomainError(MISSING_SCENARIO_PRICE);
  }
  const selling = parseAmount(sellingPriceRaw);
  if (selling.lessThanOrEqualTo(0)) {
    throw new DomainError(SELLING_PRICE_POSITIVE);
  }

  // Full-precision intermediates — rounding happens only at emission.
  const marketplaceFee = selling.times(parseAmount(assumptions.marketplaceFeeRate)).dividedBy(100);
  const paymentFee = selling.times(parseAmount(assumptions.paymentFeeRate)).dividedBy(100);
  const totalCost = parseAmount(assumptions.acquisitionCost)
    .plus(parseAmount(assumptions.shippingCost))
    .plus(marketplaceFee)
    .plus(paymentFee)
    .plus(parseAmount(assumptions.otherCosts));
  const grossProfit = selling.minus(totalCost);
  const grossMarginPercent = grossProfit.dividedBy(selling).times(100);

  return {
    scenario: scenarioName,
    sellingPrice: formatAmount(selling),
    acquisitionCost: formatAmount(parseAmount(assumptions.acquisitionCost)),
    shippingCost: formatAmount(parseAmount(assumptions.shippingCost)),
    marketplaceFeeRate: formatAmount(parseAmount(assumptions.marketplaceFeeRate)),
    marketplaceFee: formatAmount(roundAmount(marketplaceFee)),
    paymentFeeRate: formatAmount(parseAmount(assumptions.paymentFeeRate)),
    paymentFee: formatAmount(roundAmount(paymentFee)),
    otherCosts: formatAmount(parseAmount(assumptions.otherCosts)),
    totalCost: formatAmount(roundAmount(totalCost)),
    grossProfit: formatAmount(roundAmount(grossProfit)),
    grossMarginPercent: formatAmount(roundAmount(grossMarginPercent)),
  };
}

/** Calculates all three scenarios in low/base/high order; all-or-nothing. */
export function calculateScenarios(assumptions: CostAssumptions): EconomicsScenario[] {
  try {
    return SCENARIO_ORDER.map(([name, key]) => calculateScenario(name, assumptions.scenarioPrices[key], assumptions));
  } catch (error) {
    if (error instanceof MoneyError) {
      throw new DomainError(error.message);
    }
    throw error;
  }
}
